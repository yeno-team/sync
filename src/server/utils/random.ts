import crypto from 'crypto';


/**
 * Dependencies for ControllerUtility class
 */
export type RandomUtilityDependencies = {
    
}

/**
 * Utility functions to help with randomness
 */
export class RandomUtility {
    /**
     * Sets the dependencies type as a private variable of the class
     * @param dependencies 
     */
    constructor(
        private dependencies: RandomUtilityDependencies
    ) {}

    /**
     * Generates a random string by length
     * @param length 
     */
    public getRandomString(length: number) {
        return crypto.randomBytes(length).toString('hex');
    }
}