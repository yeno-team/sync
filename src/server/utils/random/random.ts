import crypto from 'crypto';


/**
 * Dependencies for ControllerUtility class
 */
export type RandomUtilityDependencies = {
    
}

/**
 * Utility functions to help with randomness
 */
export class RandomUtility {
    /**
     * Sets the dependencies type as a private variable of the class
     * @param dependencies 
     */
    constructor(
        private dependencies: RandomUtilityDependencies
    ) {}

    /**
     * Generates a random string by length
     * @param length 
     */
    public getRandomString(length: number): Promise<string> {
        return new Promise((resolve, reject) => {
            crypto.randomBytes(Math.ceil(length/2), (err, buffer) => {
                if (err) {
                    return reject(err);
                }

                return resolve(buffer.toString('hex').slice(0, length));
            });
        })
    }

    /**
     * Generate a random integer
     * @param min 
     * @param max 
     * @returns a random number
     */
    public getRandomInteger(min: number, max: number): number {
        return Math.floor(Math.random() * (max - min + 1) ) + min;
    }
}