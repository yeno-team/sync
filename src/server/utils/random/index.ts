import { RandomUtility } from "./random";

export default RandomUtility;