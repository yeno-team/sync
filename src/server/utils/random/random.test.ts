import { RandomUtility } from "./random";
const randomUtil = new RandomUtility({});

function defuse(promise) {
  promise.catch(() => {});
  return promise;
}

describe("RandomUtility Class", () => {
  beforeEach(() => {
    jest.spyOn(global.Math, "random").mockReturnValue(0.2);
  });

  afterEach(() => {
    jest.spyOn(global.Math, "random").mockRestore();
  });

  describe("#getRandomString", () => {
    it.each([1, 20, 12, 5, 32, 3])(
      "should return a string with a length of %i",
      async (length) => {
        const string = await randomUtil.getRandomString(length);
        expect(string).toHaveLength(length);
      }
    );

    it("should reject if there is an error", async () => {
      await expect(
        randomUtil.getRandomString(-2)
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        `"The value of \\"size\\" is out of range. It must be >= 0 && <= 4294967295. Received -1"`
      );
    });
  });

  describe("#getRandomInteger", () => {
    it("should return a number", () => {
      expect(
        typeof randomUtil.getRandomInteger(0, 10) === "number"
      ).toBeTruthy();
    });

    it.each([
      [0, 4],
      [5, 2983],
      [8, 23],
      [2, 500],
      [-2, 3],
    ])("should return a number in range of %i and %i", (min, max) => {
      const randomInteger = randomUtil.getRandomInteger(min, max);
      expect(randomInteger > min && randomInteger <= max);
    });

    it.each([
      [-2, -4],
      [1, 0],
      [28287, 2],
      [0.3, 0.1],
      [0, -1],
      [0, 0],
    ])(
      "should throw an error if %i is greater than or equal to %i",
      (min, max) => {
        expect(() => randomUtil.getRandomInteger(min, max)).toThrow(
          "RandomUtility: Minimum number cannot be greater than or equal to max"
        );
      }
    );
  });
});
