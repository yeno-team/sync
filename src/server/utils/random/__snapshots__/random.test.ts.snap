// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomUtility Class #getRandomString should reject if there is an size is out of range 1`] = `[RangeError: The value of "size" is out of range. It must be >= 0 && <= 4294967295. Received -1]`;
